### PR TITLE
ArmPkg, StandaloneMmPkg: optimise memory permission change

### PR DESCRIPTION
**Pull request by LeviYeoReum**
# Description

In StandaloneMm, the allowed memory permission combinations are ROX, RO,
 and RW.

At the initial stage, all regions are mapped as ROX. When drivers are loaded for relocation, the image is first mapped as RW.

However, the current mechanism introduces significant overhead when changing memory permissions. For example,
when changing permissions from ROX to RW, the process is performed in two steps:

  1. Clear the execute (X) permission (ROX → RO)
  2. Add the write (W) permission (RO → RW)

To execute step (1), the system first retrieves the permissions of the target region. This may trigger multiple SMC calls, depending on the permissions of the pages within that region.

This permission retrieval operation is redundant, as there is no need to maintain an intermediate state. Nevertheless,
it can cause additional SMC calls and coherence operations from the SPMC (e.g., TLB flushes).

Therefore, this redundant operation should be removed, and in most cases, the memory permission should be updated with a single “set memory permission” operation.

- [ ] Breaking change?
  - N/A
- [ ] Impacts security?
  - N/A
- [ ] Includes tests?
  - N/A

## How This Was Tested

Booting test on FVP_RevC with StandaloneMm.

## Integration Instructions

N/A

### CODE REVIEW

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c [hunk 1]
```diff
@@ -28,6 +28,9 @@
 #include <Library/DebugLib.h>
 #include <Library/PcdLib.h>
 
+#define EFI_PAGE_SIZE_ALIGN_UP(Size) \
```
Comment (leiflindholm) [inline ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: With an EFI prefix, this would probably be better added to UefiBaseType.h in a preceding commit. And it would be a useful addition there.
Alternatively, drop the prefix.

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c [hunk 2]
```diff
@@ -50,7 +53,9 @@ IsFfaMemoryAbiSupported (
 {
   EFI_STATUS  Status;
 
-  *Version = 0;
+  if (Version != NULL) {
```
Comment (leiflindholm) [inline ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: Unrelated to rest of change, please break out separately.

Also, *Version is not `OPTIONAL`, either for this function or for `ArmFfaLibGetVersion`, so if you add the test, `Version` being `NULL` should be treated as an error.

File: ArmPkg/Include/Library/StandaloneMmMmuLib.h [hunk 3]
```diff
@@ -8,26 +8,122 @@
 
 #pragma once
 
+/**
```
Comment (leiflindholm) [inline ArmPkg/Include/Library/StandaloneMmMmuLib.h:11]: The additions of the comment headers for existing functions is good, but unrelated to the functional change. Please break out into separate commit.

**Review by leiflindholm — CHANGES_REQUESTED** [review]

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c (hunk 1 above)
Comment (LeviYeoReum) [reply ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: Okay. I'll add this into ` UefiBaseType.h`

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c (hunk 2 above)
Comment (LeviYeoReum) [reply ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: Thanks to point out this. I'll add this with separated patch.

File: ArmPkg/Include/Library/StandaloneMmMmuLib.h (hunk 3 above)
Comment (LeviYeoReum) [reply ArmPkg/Include/Library/StandaloneMmMmuLib.h:11]: Okay. I'll separate it. Thanks!

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c (hunk 1 above)
Comment (LeviYeoReum) [reply ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: I found we can use just `ALIGN_VALUE`. So I'll remove this. Thanks!

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c [hunk 4]
```diff
@@ -400,130 +440,167 @@ ArmClearMemoryRegionNoExec (
...
-               Length,
-               &MemoryAttributes,
-               &PageCount
-               );
-    if (EFI_ERROR (Status)) {
-      break;
-    }
+  if (Length == 0) {
+    /// Nothing to do.
+    return EFI_SUCCESS;
```
Comment (kuqin12) [inline ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c:492]: Is this true? Or would it better be invalid_parameter?

File: ArmPkg/Include/Library/StandaloneMmMmuLib.h [hunk 5]
```diff
@@ -8,26 +8,122 @@
...
+  @retval EFI_UNSUPPORTED         This function is not implemented by the
+                                  callee.
+  @retval EFI_ABORTED             Message target ran into an unexpected error
+                                  and has aborted.
+  @retval EFI_ACCESS_DENIED       Access denied.
+  @retval EFI_OUT_OF_RESOURCES    Out of memory to perform operation.
+
+**/
 EFI_STATUS
 ArmClearMemoryRegionReadOnly (
```
Comment (kuqin12) [inline ArmPkg/Include/Library/StandaloneMmMmuLib.h]: The other function changes are good, but the only concern is this one: it also removes the other code permissions. That makes the functionality not suitable for this function name anymore. Would it be better to name it as "ArmSetMemoryRegionReadWrite"?

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c (hunk 4 above)
Comment (LeviYeoReum) [reply ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c:492]: Since `SectionHeader.Misc.VirtualSize` could be 0 Legnth -- like .reloc section: 

```
[8001 0] UpdateMmFoundationPeCoffPermissions: Mapping section 1 of image at 0x7004000 with RW-XN permissions
[8001 0] UpdateMmFoundationPeCoffPermissions: Section 2 of image at 0x7004000 has 0x42000040 permissions
[8001 0] UpdateMmFoundationPeCoffPermissions: Section 2 of image at 0x7004000 has .reloc name
[8001 0] UpdateMmFoundationPeCoffPermissions: Section 2 of image at 0x7004000 has 0x701A000 address
[8001 0] UpdateMmFoundationPeCoffPermissions: Section 2 of image at 0x7004000 has 0x16000 data
api_ffa_mem_perm_set page_count 0
ERROR: FFA_MEM_PERM_SET: page_count was zero
```

I think for this case, just make skip for this.

File: ArmPkg/Include/Library/StandaloneMmMmuLib.h (hunk 5 above)
Comment (LeviYeoReum) [reply ArmPkg/Include/Library/StandaloneMmMmuLib.h]: This is what I did at first with the same reason But, @samimujawar wants to keep the function name for this.
@samimujawar, Should we keep the function name?

File: MdeModulePkg/Library/ArmFfaLib/ArmFfaCommon.c [hunk 6]
```diff
@@ -184,7 +184,7 @@ EFI_STATUS
 EFIAPI
 ArmFfaLibGetVersion (
   IN  UINT32  RequestVersion,
-  OUT UINT32  *CurrentVersion
+  OUT UINT32  *CurrentVersion OPTIONAL
```
Comment (leiflindholm) [inline MdeModulePkg/Library/ArmFfaLib/ArmFfaCommon.c]: This change turns a function called "GetVersion" into something that may or may not return a version.
So either the function should be renamed, a new function only checking compatibility should be added, or the argument should remain mandatory.

Regardless, the prototype in MdePkg/Include/Library/ArmFfaLib.h should be updated to match anything that happens here.

**Review by leiflindholm — CHANGES_REQUESTED** [review]

File: MdeModulePkg/Library/ArmFfaLib/ArmFfaCommon.c (hunk 6 above)
Comment (LeviYeoReum) [reply MdeModulePkg/Library/ArmFfaLib/ArmFfaCommon.c]: TBH, I think I just want to keep ArmFfaLibGetVersion() without optional.
But, the code was written the "CurrentVersion" as if this is optional parameter.
I think it would be better to remove this check but return error in case of this parameter is NULL
and it's enough to change SetPermission's IsFfaMemoryAbiSupport.

Is it okay for you?

Comment (leiflindholm) [reply MdeModulePkg/Library/ArmFfaLib/ArmFfaCommon.c]: That works for me.

File: ArmPkg/Include/Library/StandaloneMmMmuLib.h (hunk 5 above)
Comment (LeviYeoReum) [reply ArmPkg/Include/Library/StandaloneMmMmuLib.h]: I've talked with @samimujawar, and I'll change to the function name.
Thanks!

File: MdeModulePkg/Library/ArmFfaLib/ArmFfaCommon.c [hunk 7]
```diff
@@ -190,6 +190,12 @@ ArmFfaLibGetVersion (
   EFI_STATUS    Status;
   ARM_FFA_ARGS  FfaArgs;
 
+  if (CurrentVersion == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  *CurrentVersion = 0;
```
Comment (leiflindholm) [inline MdeModulePkg/Library/ArmFfaLib/ArmFfaCommon.c]: If this zero-init is needed, I suggest it should be broken out into a separate commit.

It is not necessary for the stated change, and is an (in my opinion unnecessary) change of behaviour.

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c [hunk 8]
```diff
@@ -45,17 +45,23 @@ STATIC
...
-  *Version = 0;
+  CurrentVersion = 0;
 
   Status = ArmFfaLibGetVersion (
              ARM_FFA_CREATE_VERSION (ARM_FFA_MAJOR_VERSION, ARM_FFA_MINOR_VERSION),
-             Version
+             &CurrentVersion
              );
+
+  if (Version != NULL) {
```
Comment (leiflindholm) [inline ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c:64]: Feels wrong to update *Version before checking Status. Invert order?

We also seem to be doing a lot of setting CurrentVersion to 0 (here, and then unconditionally in ArmFfaLibGetVersion in the preceding commit) when both functions actually return a status.

**Review by leiflindholm — CHANGES_REQUESTED** [review]

File: MdeModulePkg/Library/ArmFfaLib/ArmFfaCommon.c (hunk 7 above)
Comment (LeviYeoReum) [reply MdeModulePkg/Library/ArmFfaLib/ArmFfaCommon.c]: Okay. I'll remove it anyway callee will check the return value first.

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c (hunk 8 above)
Comment (LeviYeoReum) [reply ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c:64]: Okay. I'll do it later. Thanks!

**Review by leiflindholm — APPROVED** [review]
I'm happy with the code.

Would like to see comments from @samimujawar or @ardbiesheuvel on the mappings API change itself.

File: ArmPkg/Include/Library/StandaloneMmMmuLib.h [hunk 9]
```diff
@@ -27,7 +27,7 @@
 
 **/
 EFI_STATUS
-ArmSetMemoryRegionNoExec (
+ArmSetMemoryRegionNoExecBit (
   IN  EFI_PHYSICAL_ADDRESS  BaseAddress,
```
Comment (ardbiesheuvel) [inline ArmPkg/Include/Library/StandaloneMmMmuLib.h:31]: Why is it better to put `Bit` at the end?

**Comment by ardbiesheuvel** [comment]
> I'm happy with the code.
> 
> Would like to see comments from @samimujawar or @ardbiesheuvel on the mappings API change itself.

Looks fine to me, although I don't see the point of the change that puts `Bit` at the end of a bunch of function names.

File: ArmPkg/Include/Library/StandaloneMmMmuLib.h (hunk 9 above)
Comment (LeviYeoReum) [reply ArmPkg/Include/Library/StandaloneMmMmuLib.h:31]: Well to distingush with "ArmSet/ClearMemoryRegionReadOnly()" But since the following patch changes with Perm suffix, I think it doesn't need to add Bit suffix. I'll remove it. 

Thanks!

**Comment by LeviYeoReum** [comment]
> > I'm happy with the code.
> > Would like to see comments from @samimujawar or @ardbiesheuvel on the mappings API change itself.
> 
> Looks fine to me, although I don't see the point of the change that puts `Bit` at the end of a bunch of function names.

Thanks. I'll remove `Bit` suffix.

**Review by ardbiesheuvel — APPROVED** [review]

**Review by kuqin12 — APPROVED** [review]

File: MdeModulePkg/Library/ArmFfaLib/ArmFfaCommon.c
Comment (samimujawar) [inline MdeModulePkg/Library/ArmFfaLib/ArmFfaCommon.c:183]: Document error codes?

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c [hunk 10]
```diff
@@ -309,6 +309,7 @@ ArmSetMemoryRegionNoExec (
   UINTN       Size;
   UINT32      PageCount;
 
+  Status     = EFI_SUCCESS;
```
Comment (samimujawar) [inline ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: Please add a comment why the Status is set here to success.

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c [hunk 11]
```diff
@@ -391,6 +392,7 @@ ArmClearMemoryRegionNoExec (
...
```
Comment (samimujawar) [inline ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: same comment as above.

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c [hunk 12]
```diff
@@ -460,71 +462,44 @@ ArmClearMemoryRegionNoExec (
...
-                            );
-    }
+  UINT32   PermissionRequest;
+  BOOLEAN  UseFfaAbis;
 
-    if (Length < Size) {
-      Length = Size;
-    }
+  if (Length == 0) {
+    /// Nothing to do.
```
Comment (samimujawar) [inline ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: Add explanation why length can be 0

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c [hunk 13]
```diff
@@ -542,65 +517,93 @@ ArmSetMemoryRegionReadOnly (
...
-  BOOLEAN     UseFfaAbis;
-  UINT32      Version;
-  UINTN       Size;
-  UINT32      PageCount;
+  UINT32   PermissionRequest;
+  BOOLEAN  UseFfaAbis;
 
-  UseFfaAbis = IsFfaMemoryAbiSupported (&Version);
+  if (Length == 0) {
+    /// Nothing to do.
```
Comment (samimujawar) [inline ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: same comment as above.

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c [hunk 14]
```diff
@@ -542,65 +517,93 @@ ArmSetMemoryRegionReadOnly (
...
+ArmSetMemoryRegionReadOnlyExecPerm (
+  IN  EFI_PHYSICAL_ADDRESS  BaseAddress,
+  IN  UINT64                Length
+  )
+{
+  UINT32   PermissionRequest;
+  BOOLEAN  UseFfaAbis;
+
+  if (Length == 0) {
+    /// Nothing to do.
```
Comment (samimujawar) [inline ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: same comment as previous one.

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c (hunk 12 above)
Comment (LeviYeoReum) [reply ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: JFYI: see the https://github.com/tianocore/edk2/pull/12536#discussion_r3186248202

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c (hunk 10 above)
Comment (LeviYeoReum) [reply ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: Okay. But JFYI: see the https://github.com/tianocore/edk2/pull/12536#discussion_r3186248202

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c (hunk 11 above)
Comment (LeviYeoReum) [reply ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: Done

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c (hunk 13 above)
Comment (LeviYeoReum) [reply ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: Done

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c (hunk 14 above)
Comment (LeviYeoReum) [reply ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: Done

File: ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c (hunk 10 above)
Comment (samimujawar) [reply ArmPkg/Library/StandaloneMmMmuLib/ArmMmuStandaloneMmLib.c]: I understand, but the explanation needs to be there in the code. The comment you referenced above will be lost when the PR is merged.

File: MdeModulePkg/Library/ArmFfaLib/ArmFfaCommon.c
Comment (LeviYeoReum) [reply MdeModulePkg/Library/ArmFfaLib/ArmFfaCommon.c:183]: Thanks. I'll add

**Review by samimujawar — APPROVED** [review]
